README
Convert Inline::CPP code to XS code. See the module's documentation for details.
See also 'perldoc InlineX::Cookbook' and the demos folder.

If you want to convert Inline::C code to XS use InlineX::C2XS.

Build in the usual way:

perl Makefile.PL
make test
make install

Cheers,
Rob 

CHANGES
- 0.24 ... In t/t_portability.t and t/t_prehead.t change all occurrences of '__INLINE_CPP'
           to '__INLINE_CPP_' (to distinguish the pattern being sought from the defining
           of __INLINE_CPP that was created in Inline-CPP-0.40).

- 0.23 ... "eval $VERSION;" in the created PM file - as per current recommendations
           Fix typo in Makefile.PL. (Change "Test::warn" to "Test::Warn".) Thanks Davido.

- 0.22 ... Remove remove_dup function. (Simpler to override $Inline::CPP::Config::flavor_defs.)

- 0.21 ... Add the capability of writing a portable Makefile.PL.

           Add the capability of writing a MANIFEST file (that includes all of the
           generated files).

           Add 'DIST' config option.

           Require Inline-0.50 and Inline-CPP-0.39. (Had been 0.45 and 0.25 respectively.)

- 0.20 ... Fix LIBS setting. Was clobbering any value set by Inline::CPP::validate().

           Add the demos/cpp2xs_utility demo.

- 0.19 ... Add CCFLAGSEX, BOOT_F and PRE_HEAD options

           Add cpp2xs.pl (which generates the cpp2xs script)

           If EXPORT_TAGS_ALL is set, then it's now unnecessary to set
           EXPORT_OK_ALL

- 0.17 ... Add PREREQ_PM config option

           Rewrite t_autowrap_cpp.t, t_code.t, t_cpp.t, t_hello.t, t_makefile_pl.t
           and t_src_location.t to accommodate recent (and hopefully future) changes
           to the layout of generated code.

           Remove 'expected.h', 'expected_cpp.txt', 'expected_autowrap.txt',
           'expected_autowrap_c.txt', 'expected_hello.txt' and 'expected_typemap.txt'
           as these are no longer used.

- 0.16 ... Add config option 'USE'

           Substitution regex added to t_makefile_pl.t - to cater for recent ActivePerls

           Cater for the possibility that we're including iostream (not iostream.h) in
           the test scripts

- 0.15 ... Add config options EXPORT_ALL, EXPORT_OK_ALL and EXPORT_TAGS_ALL

           Add newline at the bottom of src/test.alt (as per requirements of a valid
           C/CPP source file).

           'USING' (which is useless with Inline::CPP and InlineX::CPP2XS anyway)
           now takes either a string or array reference as its value (as does
           the Inline::C equivalent).

           'LIBS' and 'TYPEMAPS' can now optionally be assigned as a (space
           delimited) string or an array reference (same as 'INC'). Previously
           'LIBS' and 'TYPEMAPS' had to be assigned as an array reference.

           'INC' (in generated Makefile.PL) no longer automatically includes the cwd.

           'TYPEMAPS' (in generated Makefile.PL) no longer automatically specifies
           the standard perl typemap. (Doing so was annoying and, afaict, unnecessary.)

           Add Inline-0.45 to PREREQ_PM.

- 0.13 ... Change the test for the locatability of a specified typemap from a
           '-e' test to a '-f' test.

           Add CPP2XS-Cookbook.pod.

           Minor change to WRITE_PM.

           The CPP source code can now be alternatively provided by either a 
           'CODE' or 'SRC_LOCATION' key. 

- 0.12 ... Provide access to ParseRegExp.pm. (Untested, because ParseRegExp.pm
           is broken - see http://rt.cpan.org/Ticket/Display.html?id=6327 for
           patches to ParseRegExp.pm)

           Can now write a .pm file for you as well

- 0.11 ... Can now provide the optional "4th" (config options) argument
           without *also* having to provide a "3rd" (build directory) argument.
           ie If the third argument is a hash reference, it's assumed to contain
           config options - otherwise it sets the build directory.
 
           Check that only valid config options have been passed via the "config
           options" argument.

           Catch the error (and die) if a non-existent typemap is specified
           in the TYPEMAPS config option.
    
           If the specified build_dir does not exist, die(). Previously, a warning was
           issued and the files written to the current working directory.

- 0.10 ... Module is now called InlineX::CPP2XS (previously Inline::CPP2XS, which is now
           deprecated).
           Now has the same capabilities as its C counterpart (InlineX::C2XS-0.10) - 
           ie can also (optionally) autogenerate the Makefile.PL in addition to the other
           files it has always been capable of generating (namely the .xs file, INLINE.h
           and CPP.map).

- 0.07 ... Inline::CPP2XS born (previously part of Inline::C2XS)


demos/cpp2xs_utility/src/FastSieve.cpp
/* This code copied from David Oswald's FastSieve.pm (Math-Prime-FastSieve-0.07)
 * "Sieve" class renamed to "_Sieve" to accommodate --write_export_ok_all option
 */

#include <vector>
using namespace std;

/* Class _Sieve below.  Perl sees it as a class named
 * "Math::Prime::FastSieve::Sieve".  The constructor is mapped to
 * "new()" within Perl, and the destructor to "DESTROY()".  All other
 * methods are mapped with the same name as declared in the class.
 *
 * Therefore, Perl sees this class approximately like this:
 *
 * package Math::Prime::FastSieve;
 *
 * sub new {
 *     my $class = shift;
 *     my $n     = shift;
 *     my $self = bless {}, $class;
 *     $self->{max_n} = n;
 *     $self->{num_primes} = 0;
 *     # Build the sieve here...
 *     # I won't bother translating it to Perl.
 *     $self->{sieve} = $primes;  // A reference to a bit vector.
 *     return $self;
 *  }
 *
 */


class _Sieve
{
    public:
        _Sieve ( int n ); // Constructor. Perl sees "new()".
        ~_Sieve(       ); // Destructor. Seen as "DESTROY()".
        bool isprime( int n ); // Test if n is prime.
        SV*  primes ( int n ); // Return all primes in an aref.
        unsigned int  nearest_le ( int n ); // Return nearest prime <= n.
        unsigned int  nearest_ge ( int n ); // Return nearest prime >= n.
        unsigned int  nth_prime  ( int n ); // Return the nth prime.
        unsigned int  count_sieve(       ); // Retrn count of primes in sieve.
        unsigned int  count_le   ( int n ); // Return number of primes <= n.
        SV*  ranged_primes( int lower, int upper );
                  // Return all primes where "lower <= primes <= upper".
    private:
        std::vector<bool>::size_type    max_n;
        unsigned int                    num_primes;
        vector<bool>* sieve;
};


// Set up a primes sieve of 0 .. n inclusive.
_Sieve::_Sieve( int n )
{
    std::vector<bool>* primes = new std::vector<bool>( n + 1, 0 );
    num_primes = 0;
    if( n < 0 ) // Trap negative n's before we start wielding unsigned ints.
        max_n = 0;
    else
    {
        max_n      = n;
        for( std::vector<bool>::size_type i = 3; i * i <= n; i+=2 )
            if( ! (*primes)[i] )
                for( std::vector<bool>::size_type k, j=i; (k=i*j) <= n; j++ )
                    (*primes)[k] = 1;
    }
    sieve = primes;
}


// Deallocate memory for primes sieve.
_Sieve::~_Sieve() {
    delete sieve;
}


// Yes or no: Is the number a prime?  Must be within the range of
// 0 through max_n (the upper limit set by the constructor).
bool _Sieve::isprime( int n )
{
    if( n < 2 || n > max_n )  return false; // Bounds checking.
    if( n == 2 )              return true;  // 2 is prime.
    if( ! ( n % 2 ) )         return false; // No other evens are prime.
    if( ! ( (*sieve)[n] ) )   return true;  // 0 bit signifies prime.
    return false;                           // default: not prime.
}


// Return a reference to an array containing the list of all primes
// less than or equal to n.  n must be within the range set in the
// constructor.
SV* _Sieve::primes( int n )
{
    AV* av = newAV();
    if( n < 2 || n > max_n ) // Logical short circuit order is significant
                             // since we're about to wield unsigned ints.
        return newRV_noinc( (SV*) av );
    av_push( av, newSVuv( 2U ) );
    num_primes = 1; // Count 2; it's prime.
    for( std::vector<bool>::size_type i = 3; i <= n; i += 2 )
        if( ! (*sieve)[i] )
            av_push( av, newSVuv( i ) );
    return newRV_noinc( (SV*) av );
}

SV* _Sieve::ranged_primes( int lower, int upper )
{
    AV* av = newAV();
    if(
        upper > max_n ||        // upper exceeds upper bound.
        lower > max_n ||        // lower exceeds upper bound.
        upper < 2     ||        // No possible primes.
        lower < 0     ||        // lower underruns bounds.
        lower > upper ||        // zero-width range.
        ( lower == upper && lower > 2 && !( lower % 2 ) ) // Even.
    )
        return newRV_noinc( (SV*) av );  // No primes possible.
    if( lower <= 2 && upper >= 2 )
        av_push( av, newSVuv( 2U ) );    // Lower limit needs to be odd
    if( lower < 3 ) lower = 3;           // Lower limit cannot < 3.
    if( ( upper - lower ) > 0 && ! ( lower % 2 ) ) lower++;
    for( std::vector<bool>::size_type i = lower; i <= upper; i += 2 )
        if( ! (*sieve)[i] )
            av_push( av, newSVuv( i ) );
    return newRV_noinc( (SV*) av );
}


// Find the nearest prime less than or equal to n.  Very fast.
unsigned int _Sieve::nearest_le( int n )
{
    // Remember that order of testing is significant; we have to
    // disqualify negative numbers before we do comparisons against
    // unsigned ints.
    if( n < 2 || n > max_n ) return 0; // Bounds checking.
    if( n == 2 ) return 2U;            // 2 is the only even prime.
    if( ! ( n % 2 ) ) n--;             // Result has to be odd.
    std::vector<bool>::size_type n_index = n;
    while( n_index > 2 )
    {
        if( ! (*sieve)[n_index] ) return n_index;
        n_index -= 2;  // Only test odds.
    }
    return 0; // We should never get here.
}


// Find the nearest prime greater than or equal to n.  Very fast.
unsigned int _Sieve::nearest_ge( int n )
{
    // Order of bounds tests IS significant.
    // Because max_n is unsigned, testing "n > max_n" for values where
    // n is negative results in n being treated as a real big unsigned value.
    // Thus we MUST handle negatives before testing max_n.
    if( n <= 2 ) return 2U;              // 2 is only even prime.
    if( n > max_n ) return 0U;           // Bounds checking.
    std::vector<bool>::size_type n_idx = n;
    if( ! ( n_idx % 2 ) ) n_idx++;              // Start with an odd number.
    while( n_idx <= max_n )
    {
        if( ! (*sieve)[n_idx] ) return n_idx;
        n_idx+=2; // Only test odds.
    }
    return 0U;   // We've run out of sieve to test.
}


// Since we're only storing the sieve (not the primes list), this is a
// linear time operation: O(n).
unsigned int _Sieve::nth_prime( int n )
{
    if( n <  1     ) return 0; // Why would anyone want the 0th prime?
    if( n >  max_n ) return 0; // There can't be more primes than sieve.
    if( n == 1     ) return 2; // We have to handle the only even prime.
    unsigned int count = 1;
    for( std::vector<bool>::size_type i = 3; i <= max_n; i += 2 )
    {
        if( ! (* sieve)[i] ) count++;
        if( count == n ) return i;
    }
    return 0;
}


// Return the number of primes in the sieve.  Once results are
// calculated, they're cached.  First time through is O(n).
unsigned int _Sieve::count_sieve ()
{
    if( num_primes > 0 ) return num_primes;
    num_primes = this->count_le( max_n );
    return num_primes;
}


// Return the number of primes less than or equal to n.  If n == max_n
// the data member num_primes will be set.
unsigned int _Sieve::count_le( int n )
{
    if( n <= 1 || n > max_n ) return 0;
    unsigned int count = 1;      // 2 is prime. Count it.
    for( std::vector<bool>::size_type i = 3; i <= n; i+=2 )
        if( !(*sieve)[i] ) count++;
    if( n == max_n && num_primes == 0 ) num_primes = count;
    return count;
}


// ---------------- For export: Not part of _Sieve class ----------------

/* Sieve of Eratosthenes.  Return a reference to an array containing all
 * prime numbers less than or equal to search_to.
 */


SV* primes( int search_to )
{
    AV* av = newAV();
    if( search_to < 2 )
        return newRV_noinc( (SV*) av ); // Return an empty list ref.
    av_push( av, newSVuv( 2U ) );
    std::vector<bool> primes( search_to + 1, 0 );
    for( std::vector<bool>::size_type i = 3; i * i <= search_to; i+=2 )
        if( ! primes[i] )
            for
            (
                    std::vector<bool>::size_type k, j = i;
                    ( k = i * j ) <= search_to;
                    j++
            )
                primes[ k ] = 1;
    for( std::vector<bool>::size_type i = 3; i <= search_to; i+=2 )
        if( ! primes[i] )
            av_push( av, newSVuv( static_cast<unsigned int>( i ) ) );
    return newRV_noinc( (SV*) av );
}





Makefile.PL
use ExtUtils::MakeMaker;

my $file = 'cpp2xs';

open RD, '<', 'cpp_script.in' or die "Can't open cpp_script.in for reading: $!";
open OUT, '>', $file or die "Can't open $file for writing: $!";

print OUT <<"!GROK!THIS!";
$Config::Config{'startperl'}
    eval 'exec perl -S \$0 "\$@"'
        if 0;
!GROK!THIS!

while(<RD>) { print OUT $_ }

close RD or die "Can't close cpp_script.in after reading: $!";
close OUT or die "Can't close $file after writing: $!";

chmod 0555, $file;

my $twv = $] > 5.013 ? 0.23 : 0.21;
$twv = 0.24 if $] > 5.015;

my %options =
  (
  NAME     => 'InlineX::CPP2XS',
  PREREQ_PM => {
                'Inline'       => 0.50,
                'Inline::CPP'  => 0.39,
                'Test::Warn'   => $twv,
               },
  VERSION_FROM => 'CPP2XS.pm',

  EXE_FILES => ['cpp2xs'],
  clean => { FILES => 'cpp2xs' },
  );

WriteMakefile(%options);




